File: frontend-panel/src/lib/audit.ts
import type { TFunction } from "i18next";
import type {
	AuditAction,
	AuditEntityType,
	AuditLogEntry,
	AuditPresentationMessage,
} from "@/types/api";

export const AUDIT_ENTITY_TYPE_FILTER_VALUES = [
	"system",
	"system_config",
	"user",
	"auth_session",
	"passkey",
	"external_auth_provider",
	"external_auth_identity",
	"api_token",
	"mail",
	"task",
	"minecraft_profile",
	"minecraft_texture",
	"yggdrasil_token",
	"yggdrasil_session",
] as const satisfies readonly AuditEntityType[];

export function isAuditEntityType(value: string): value is AuditEntityType {
	type MissingAuditEntityType = Exclude<
		AuditEntityType,
		(typeof AUDIT_ENTITY_TYPE_FILTER_VALUES)[number]
	>;
	const filterValuesCoverOpenApi: MissingAuditEntityType extends never
		? true
		: never = true;
	return (
		filterValuesCoverOpenApi &&
		AUDIT_ENTITY_TYPE_FILTER_VALUES.includes(value as AuditEntityType)
	);
}

function resolveAuditTranslation(t: TFunction, key: string, fallback?: string) {
	const translated = t(key, { defaultValue: key });
	return translated === key ? fallback : translated;
}

export function formatAuditAction(t: TFunction, action: AuditAction | string) {
	const value = String(action);
	return (
		resolveAuditTranslation(t, `admin.audit.action.${value}`) ??
		humanizeCode(value)
	);
}

export function formatAuditEntityType(
	t: TFunction,
	entityType: string | null | undefined,
) {
	if (!entityType) {
		return "---";
	}

	return (
		resolveAuditTranslation(t, `admin.audit.entity.${entityType}`) ??
		humanizeCode(entityType)
	);
}

type AuditPresentationEntry = Pick<AuditLogEntry, "action" | "presentation">;

function isRecord(value: unknown): value is Record<string, unknown> {
	return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function presentationParams(
	message: AuditPresentationMessage | undefined | null,
) {
	return isRecord(message?.params) ? message.params : {};
}

function stringParam(
	params: Record<string, unknown>,
	key: string,
): string | undefined {
	const value = params[key];
	if (typeof value === "string") {
		return value;
	}
	if (typeof value === "number" || typeof value === "boolean") {
		return String(value);
	}
	if (Array.isArray(value)) {
		const text = value.map((item) => String(item)).join(", ");
		return text || undefined;
	}
	return undefined;
}

function humanizeCode(value: string) {
	const words = value.replaceAll("-", "_").split("_").filter(Boolean);
	if (words.length === 0) {
		return value;
	}
	const text = words.join(" ");
	return text.charAt(0).toUpperCase() + text.slice(1);
}

function formatAuditPresentationMessage(
	t: TFunction,
	message: AuditPresentationMessage | undefined | null,
	prefer: "action" | "presentation" | "target",
	fallback?: string,
) {
	if (!message?.code) {
		return fallback;
	}

	const params = presentationParams(message);
	if (prefer === "action") {
		const actionLabel = resolveAuditTranslation(
			t,
			`admin.audit.action.${message.code}`,
		);
		if (actionLabel) {
			return actionLabel;
		}
	}

	const direct = resolveAuditTranslation(
		t,
		`admin.audit.presentation.${message.code}`,
	);
	if (direct) {
		return t(`admin.audit.presentation.${message.code}`, {
			defaultValue: direct,
			...params,
		});
	}

	const actionLabel = resolveAuditTranslation(
		t,
		`admin.audit.action.${message.code}`,
	);
	if (actionLabel) {
		return actionLabel;
	}

	const entityLabel = resolveAuditTranslation(
		t,
		`admin.audit.entity.${message.code}`,
	);
	if (entityLabel) {
		const name = stringParam(params, "name");
		const id = stringParam(params, "id");
		if (prefer === "target" && name) {
			return `${name} · ${entityLabel}`;
		}
		if (prefer === "target" && id) {
			return `${entityLabel} #${id}`;
		}
		return entityLabel;
	}

	return fallback;
}

export function formatAuditSummary(
	t: TFunction,
	entry: AuditPresentationEntry,
) {
	return (
		formatAuditPresentationMessage(t, entry.presentation?.summary, "action") ??
		formatAuditAction(t, entry.action)
	);
}

export function formatAuditTarget(
	t: TFunction,
	entry: Pick<
		AuditLogEntry,
		"entity_id" | "entity_name" | "entity_type" | "presentation"
	>,
) {
	const fallback = [
		entry.entity_name ?? null,
		formatAuditEntityType(t, entry.entity_type),
		entry.entity_id == null ? null : `#${entry.entity_id}`,
	]
		.filter((value): value is string => Boolean(value))
		.join(" · ");
	return (
		formatAuditPresentationMessage(
			t,
			entry.presentation?.target,
			"target",
			fallback,
		) ?? fallback
	);
}

export function formatAuditTargetType(
	t: TFunction,
	entry: Pick<AuditLogEntry, "entity_type" | "presentation">,
) {
	const targetCode = entry.presentation?.target?.code;
	if (targetCode) {
		const translatedTargetType = resolveAuditTranslation(
			t,
			`admin.audit.entity.${targetCode}`,
		);
		if (translatedTargetType) {
			return translatedTargetType;
		}
	}
	return formatAuditEntityType(t, entry.entity_type);
}

export function formatAuditDetail(t: TFunction, entry: AuditPresentationEntry) {
	return formatAuditPresentationMessage(
		t,
		entry.presentation?.detail,
		"presentation",
	);
}

type AuditActionTone = "danger" | "success" | "info" | "warning";

const AUDIT_ACTION_TONES = {
	admin_delete_config: "danger",
	admin_delete_external_auth_provider: "danger",
	admin_disable_user: "danger",
	admin_revoke_user_sessions: "danger",
	config_delete: "danger",
	external_auth_provider_delete: "danger",
	mail_delivery_failed: "danger",
	minecraft_profile_delete: "danger",
	minecraft_texture_delete: "danger",
	server_shutdown: "danger",
	user_revoke_other_sessions: "danger",
	user_revoke_session: "danger",
	yggdrasil_invalidate_token: "danger",
	yggdrasil_signout: "danger",

	mail_send: "success",
	minecraft_profile_create: "success",
	minecraft_texture_upload: "success",
	system_setup: "success",
	user_register: "success",

	admin_create_external_auth_provider: "info",
	admin_create_user: "info",
	admin_update_external_auth_provider: "info",
	admin_update_user: "info",
	config_action_execute: "info",
	config_update: "info",
	external_auth_provider_create: "info",
	external_auth_provider_update: "info",
	minecraft_texture_bind: "info",
	task_retry: "info",
	user_change_password: "info",
	user_update_profile: "info",

	admin_test_external_auth_provider: "warning",
	user_external_auth_link: "warning",
	user_external_auth_login: "warning",
	user_external_auth_unlink: "warning",
	user_login: "warning",
	user_logout: "warning",
	user_refresh_token: "warning",
	yggdrasil_authenticate: "warning",
	yggdrasil_join_server: "warning",
	yggdrasil_refresh_token: "warning",
} as const satisfies Partial<Record<AuditAction, AuditActionTone>>;

const AUDIT_ACTION_TONE_CLASSES = {
	danger:
		"border-red-200 bg-red-50 text-red-700 dark:border-red-900 dark:bg-red-950/60 dark:text-red-300",
	info: "border-sky-200 bg-sky-50 text-sky-700 dark:border-sky-900 dark:bg-sky-950/60 dark:text-sky-300",
	success:
		"border-emerald-200 bg-emerald-50 text-emerald-700 dark:border-emerald-900 dark:bg-emerald-950/60 dark:text-emerald-300",
	warning:
		"border-amber-200 bg-amber-50 text-amber-700 dark:border-amber-900 dark:bg-amber-950/60 dark:text-amber-300",
} as const satisfies Record<AuditActionTone, string>;

export function getAuditActionBadgeClass(action: AuditAction | string) {
	const tone =
		typeof action === "string" &&
		Object.hasOwn(AUDIT_ACTION_TONES, action) &&
		AUDIT_ACTION_TONES[action as keyof typeof AUDIT_ACTION_TONES];
	if (tone) {
		return AUDIT_ACTION_TONE_CLASSES[tone];
	}
	return "border-border bg-muted/30 text-muted-foreground";
}

File: src/types/audit.rs
use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};
use std::fmt;
#[cfg(all(debug_assertions, feature = "openapi"))]
use utoipa::ToSchema;

macro_rules! define_audit_actions {
    ($($variant:ident => $name:literal),+ $(,)?) => {
        #[derive(Debug, Clone, Copy, PartialEq, Eq, EnumIter, DeriveActiveEnum, Serialize, Deserialize)]
        #[cfg_attr(all(debug_assertions, feature = "openapi"), derive(ToSchema))]
        #[sea_orm(rs_type = "String", db_type = "String(StringLen::N(64))")]
        #[serde(rename_all = "snake_case")]
        pub enum AuditAction {
            $(
                #[sea_orm(string_value = $name)]
                #[serde(rename = $name)]
                $variant,
            )+
        }

        impl AuditAction {
            pub const COUNT: usize = <[()]>::len(&[$(define_audit_actions!(@unit $variant)),+]);
            pub const ALL: [Self; Self::COUNT] = [$(Self::$variant,)+];

            pub const fn as_str(self) -> &'static str {
                match self {
                    $(Self::$variant => $name,)+
                }
            }

            pub fn from_str_name(value: &str) -> Option<Self> {
                match value {
                    $($name => Some(Self::$variant),)+
                    _ => None,
                }
            }

            pub fn index(self) -> usize {
                Self::ALL
                    .iter()
                    .position(|action| *action == self)
                    .expect("audit action should be present in AuditAction::ALL")
            }
        }
    };
    (@unit $variant:ident) => { () };
}

define_audit_actions! {
    SystemSetup => "system_setup",
    ServerStart => "server_start",
    ServerShutdown => "server_shutdown",
    ConfigUpdate => "config_update",
    ConfigDelete => "config_delete",
    ConfigActionExecute => "config_action_execute",
    UserRegister => "user_register",
    UserLogin => "user_login",
    UserLogout => "user_logout",
    UserRefreshToken => "user_refresh_token",
    UserRevokeSession => "user_revoke_session",
    UserRevokeOtherSessions => "user_revoke_other_sessions",
    UserChangePassword => "user_change_password",
    UserConfirmRegistration => "user_confirm_registration",
    UserRequestEmailChange => "user_request_email_change",
    UserResendEmailChange => "user_resend_email_change",
    UserConfirmEmailChange => "user_confirm_email_change",
    UserRequestPasswordReset => "user_request_password_reset",
    UserConfirmPasswordReset => "user_confirm_password_reset",
    UserUpdateProfile => "user_update_profile",
    UserPasskeyRegister => "user_passkey_register",
    UserPasskeyRename => "user_passkey_rename",
    UserPasskeyDelete => "user_passkey_delete",
    UserPasskeyLogin => "user_passkey_login",
    AdminCreateUser => "admin_create_user",
    AdminUpdateUser => "admin_update_user",
    AdminDisableUser => "admin_disable_user",
    AdminDeleteUser => "admin_delete_user",
    AdminCreateInvitation => "admin_create_invitation",
    AdminRevokeInvitation => "admin_revoke_invitation",
    AdminRevokeUserSessions => "admin_revoke_user_sessions",
    AdminDeleteConfig => "admin_delete_config",
    AdminCleanupTasks => "admin_cleanup_tasks",
    TaskRetry => "task_retry",
    AdminCreateExternalAuthProvider => "admin_create_external_auth_provider",
    AdminUpdateExternalAuthProvider => "admin_update_external_auth_provider",
    AdminDeleteExternalAuthProvider => "admin_delete_external_auth_provider",
    AdminTestExternalAuthProvider => "admin_test_external_auth_provider",
    MailSend => "mail_send",
    MailDeliveryFailed => "mail_delivery_failed",
    ExternalAuthProviderCreate => "external_auth_provider_create",
    ExternalAuthProviderUpdate => "external_auth_provider_update",
    ExternalAuthProviderDelete => "external_auth_provider_delete",
    UserExternalAuthLogin => "user_external_auth_login",
    UserExternalAuthLink => "user_external_auth_link",
    UserExternalAuthUnlink => "user_external_auth_unlink",
    MinecraftProfileCreate => "minecraft_profile_create",
    MinecraftProfileRename => "minecraft_profile_rename",
    MinecraftProfileDelete => "minecraft_profile_delete",
    MinecraftTextureUpload => "minecraft_texture_upload",
    MinecraftTextureBind => "minecraft_texture_bind",
    MinecraftTextureDelete => "minecraft_texture_delete",
    MinecraftTextureLibrarySubmit => "minecraft_texture_library_submit",
    MinecraftTextureLibraryWithdraw => "minecraft_texture_library_withdraw",
    MinecraftTextureLibraryApprove => "minecraft_texture_library_approve",
    MinecraftTextureLibraryReject => "minecraft_texture_library_reject",
    MinecraftTextureLibraryUnpublish => "minecraft_texture_library_unpublish",
    MinecraftTextureReportCreate => "minecraft_texture_report_create",
    MinecraftTextureReportAccept => "minecraft_texture_report_accept",
    MinecraftTextureReportReject => "minecraft_texture_report_reject",
    YggdrasilAuthenticate => "yggdrasil_authenticate",
    YggdrasilRefreshToken => "yggdrasil_refresh_token",
    YggdrasilInvalidateToken => "yggdrasil_invalidate_token",
    YggdrasilSignout => "yggdrasil_signout",
    YggdrasilJoinServer => "yggdrasil_join_server",
}

impl AsRef<str> for AuditAction {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl AuditAction {
    pub const fn group(self) -> &'static str {
        match self {
            Self::SystemSetup | Self::ServerStart | Self::ServerShutdown => "system",
            Self::ConfigUpdate
            | Self::ConfigDelete
            | Self::ConfigActionExecute
            | Self::AdminDeleteConfig => "config",
            Self::AdminCleanupTasks | Self::TaskRetry => "task",
            Self::UserRegister
            | Self::UserLogin
            | Self::UserLogout
            | Self::UserRefreshToken
            | Self::UserRevokeSession
            | Self::UserRevokeOtherSessions
            | Self::UserChangePassword
            | Self::UserConfirmRegistration
            | Self::UserRequestEmailChange
            | Self::UserResendEmailChange
            | Self::UserConfirmEmailChange
            | Self::UserRequestPasswordReset
            | Self::UserConfirmPasswordReset
            | Self::UserUpdateProfile
            | Self::UserPasskeyRegister
            | Self::UserPasskeyRename
            | Self::UserPasskeyDelete
            | Self::UserPasskeyLogin
            | Self::MinecraftProfileCreate
            | Self::MinecraftProfileRename
            | Self::MinecraftProfileDelete
            | Self::MinecraftTextureUpload
            | Self::MinecraftTextureBind
            | Self::MinecraftTextureDelete
            | Self::MinecraftTextureLibrarySubmit
            | Self::MinecraftTextureLibraryWithdraw
            | Self::MinecraftTextureReportCreate => "user",
            Self::YggdrasilAuthenticate
            | Self::YggdrasilRefreshToken
            | Self::YggdrasilInvalidateToken
            | Self::YggdrasilSignout
            | Self::YggdrasilJoinServer => "yggdrasil",
            Self::AdminCreateUser
            | Self::AdminUpdateUser
            | Self::AdminDisableUser
            | Self::AdminDeleteUser
            | Self::AdminCreateInvitation
            | Self::AdminRevokeInvitation
            | Self::AdminRevokeUserSessions => "admin",
            Self::MinecraftTextureLibraryApprove
            | Self::MinecraftTextureLibraryReject
            | Self::MinecraftTextureLibraryUnpublish
            | Self::MinecraftTextureReportAccept
            | Self::MinecraftTextureReportReject => "admin",
            Self::MailSend | Self::MailDeliveryFailed => "mail",
            Self::AdminCreateExternalAuthProvider
            | Self::AdminUpdateExternalAuthProvider
            | Self::AdminDeleteExternalAuthProvider
            | Self::AdminTestExternalAuthProvider
            | Self::ExternalAuthProviderCreate
            | Self::ExternalAuthProviderUpdate
            | Self::ExternalAuthProviderDelete
            | Self::UserExternalAuthLogin
            | Self::UserExternalAuthLink
            | Self::UserExternalAuthUnlink => "external_auth",
        }
    }
}

impl fmt::Display for AuditAction {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(all(debug_assertions, feature = "openapi"), derive(ToSchema))]
#[serde(rename_all = "snake_case")]
pub enum AuditEntityType {
    System,
    SystemConfig,
    User,
    Invitation,
    AuthSession,
    Passkey,
    ExternalAuthProvider,
    ExternalAuthIdentity,
    ApiToken,
    Mail,
    Task,
    MinecraftProfile,
    MinecraftTexture,
    YggdrasilToken,
    YggdrasilSession,
}

impl AuditEntityType {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::System => "system",
            Self::SystemConfig => "system_config",
            Self::User => "user",
            Self::Invitation => "invitation",
            Self::AuthSession => "auth_session",
            Self::Passkey => "passkey",
            Self::ExternalAuthProvider => "external_auth_provider",
            Self::ExternalAuthIdentity => "external_auth_identity",
            Self::ApiToken => "api_token",
            Self::Mail => "mail",
            Self::Task => "task",
            Self::MinecraftProfile => "minecraft_profile",
            Self::MinecraftTexture => "minecraft_texture",
            Self::YggdrasilToken => "yggdrasil_token",
            Self::YggdrasilSession => "yggdrasil_session",
        }
    }

    pub fn from_str_name(value: &str) -> Option<Self> {
        match value {
            "system" => Some(Self::System),
            "system_config" => Some(Self::SystemConfig),
            "user" => Some(Self::User),
            "invitation" => Some(Self::Invitation),
            "auth_session" => Some(Self::AuthSession),
            "passkey" => Some(Self::Passkey),
            "external_auth_provider" => Some(Self::ExternalAuthProvider),
            "external_auth_identity" => Some(Self::ExternalAuthIdentity),
            "api_token" => Some(Self::ApiToken),
            "mail" => Some(Self::Mail),
            "task" => Some(Self::Task),
            "minecraft_profile" => Some(Self::MinecraftProfile),
            "minecraft_texture" => Some(Self::MinecraftTexture),
            "yggdrasil_token" => Some(Self::YggdrasilToken),
            "yggdrasil_session" => Some(Self::YggdrasilSession),
            _ => None,
        }
    }
}

impl AsRef<str> for AuditEntityType {
    fn as_ref(&self) -> &str {
        self.as_str()
    }
}

impl fmt::Display for AuditEntityType {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}
